File: backend/app/db/models/training.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy import CheckConstraint, DateTime, Enum as SqlEnum, ForeignKey, Integer, String, UniqueConstraint, Uuid
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.db.models.common import utc_now


class UnresolvedReviewStatus(str, Enum):
    OPEN = "open"
    UNDER_REVIEW = "under_review"
    MAPPED = "mapped"
    DISMISSED = "dismissed"
    NOT_ACTIONABLE = "not_actionable"


class MappingStatus(str, Enum):
    SUGGESTED = "suggested"
    DRAFT = "draft"
    TESTING = "testing"
    APPROVED = "approved"
    PUBLISHED = "published"
    SUPERSEDED = "superseded"
    REJECTED = "rejected"


class MappingOrigin(str, Enum):
    BUILT_IN = "built_in"
    ADMINISTRATOR = "administrator"
    AI_ASSISTED = "ai_assisted"


class ValidationRunStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    PASSED = "passed"
    FAILED = "failed"


class UnresolvedBlock(Base):
    __tablename__ = "unresolved_blocks"
    __table_args__ = (
        CheckConstraint("char_length(raw_text) <= 2048", name="ck_unresolved_blocks_raw_text_bounded"),
        CheckConstraint("char_length(surrounding_context) <= 4096", name="ck_unresolved_blocks_context_bounded"),
        CheckConstraint("char_length(fingerprint) = 64", name="ck_unresolved_blocks_fingerprint"),
        UniqueConstraint("audit_id", "fingerprint"),
    )

    unresolved_block_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    organization_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("organizations.organization_id", ondelete="RESTRICT"), nullable=False, index=True)
    audit_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("audits.audit_id", ondelete="RESTRICT"), nullable=False, index=True)
    device_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("devices.device_id", ondelete="RESTRICT"), nullable=False, index=True)
    snapshot_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("snapshots.snapshot_id", ondelete="RESTRICT"), nullable=False, index=True)
    profile_id: Mapped[str | None] = mapped_column(String(255))
    profile_version_id: Mapped[str | None] = mapped_column(String(255))
    source_ir_node_ids: Mapped[list[str]] = mapped_column(JSONB, nullable=False, default=list)
    evidence_refs: Mapped[list[dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)
    raw_text: Mapped[str] = mapped_column(String(2048), nullable=False)
    surrounding_context: Mapped[str] = mapped_column(String(4096), nullable=False, default="")
    unknown_reason: Mapped[str] = mapped_column(String(128), nullable=False)
    candidate_field_ids: Mapped[list[str]] = mapped_column(JSONB, nullable=False, default=list)
    affected_rule_ids: Mapped[list[str]] = mapped_column(JSONB, nullable=False, default=list)
    fingerprint: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    occurrence: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    review_status: Mapped[UnresolvedReviewStatus] = mapped_column(SqlEnum(UnresolvedReviewStatus, name="ck_unresolved_blocks_review_status", native_enum=False, create_constraint=True, validate_strings=True, values_callable=lambda enum: [item.value for item in enum], length=32), nullable=False, default=UnresolvedReviewStatus.OPEN, server_default=UnresolvedReviewStatus.OPEN.value, index=True)
    assigned_mapping_version_id: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("mapping_versions.mapping_version_id", ondelete="RESTRICT"), index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now, onupdate=utc_now)
    schema_version: Mapped[str] = mapped_column(String(32), nullable=False, default="1.0.0", server_default="1.0.0")


class MappingVersion(Base):
    __tablename__ = "mapping_versions"
    __table_args__ = (
        UniqueConstraint("mapping_id", "version"),
        CheckConstraint("version > 0", name="ck_mapping_versions_version_positive"),
    )

    mapping_version_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    mapping_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), nullable=False, index=True)
    organization_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("organizations.organization_id", ondelete="RESTRICT"), nullable=False, index=True)
    mapping_key: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    previous_mapping_version_id: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("mapping_versions.mapping_version_id", ondelete="RESTRICT"))
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(String(2048), nullable=False)
    status: Mapped[MappingStatus] = mapped_column(SqlEnum(MappingStatus, name="ck_mapping_versions_status", native_enum=False, create_constraint=True, validate_strings=True, values_callable=lambda enum: [item.value for item in enum], length=32), nullable=False, index=True)
    profile_applicability: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    structural_match: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    target_field_id: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    value_extraction: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    unit_conversion: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    scope_resolution: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    negation_behavior: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    removal_behavior: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    default_behavior: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    examples: Mapped[list[dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)
    validation_results: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    origin: Mapped[MappingOrigin] = mapped_column(SqlEnum(MappingOrigin, name="ck_mapping_versions_origin", native_enum=False, create_constraint=True, validate_strings=True, values_callable=lambda enum: [item.value for item in enum], length=32), nullable=False)
    ai_suggestion_metadata: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    created_by: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("users.user_id", ondelete="RESTRICT"))
    approved_by: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("users.user_id", ondelete="RESTRICT"))
    knowledge_pack_version_id: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("knowledge_pack_versions.knowledge_pack_version_id", ondelete="RESTRICT"), index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    schema_version: Mapped[str] = mapped_column(String(32), nullable=False, default="1.0.0", server_default="1.0.0")


class MappingValidationRun(Base):
    __tablename__ = "mapping_validation_runs"
    validation_run_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    organization_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("organizations.organization_id", ondelete="RESTRICT"), nullable=False, index=True)
    mapping_version_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("mapping_versions.mapping_version_id", ondelete="RESTRICT"), nullable=False, index=True)
    status: Mapped[ValidationRunStatus] = mapped_column(SqlEnum(ValidationRunStatus, name="ck_mapping_validation_runs_status", native_enum=False, create_constraint=True, validate_strings=True, values_callable=lambda enum: [item.value for item in enum], length=16), nullable=False, default=ValidationRunStatus.PENDING, server_default=ValidationRunStatus.PENDING.value)
    results: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    schema_version: Mapped[str] = mapped_column(String(32), nullable=False, default="1.0.0", server_default="1.0.0")


class KnowledgePackRecord(Base):
    __tablename__ = "knowledge_packs"
    knowledge_pack_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    organization_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("organizations.organization_id", ondelete="RESTRICT"), nullable=False, index=True)
    pack_key: Mapped[str] = mapped_column(String(255), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now)
    schema_version: Mapped[str] = mapped_column(String(32), nullable=False, default="1.0.0", server_default="1.0.0")
    __table_args__ = (UniqueConstraint("organization_id", "pack_key"),)


class KnowledgePackVersionRecord(Base):
    __tablename__ = "knowledge_pack_versions"
    __table_args__ = (UniqueConstraint("knowledge_pack_id", "version"), CheckConstraint("version > 0", name="ck_knowledge_pack_versions_version_positive"))
    knowledge_pack_version_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    knowledge_pack_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("knowledge_packs.knowledge_pack_id", ondelete="RESTRICT"), nullable=False, index=True)
    organization_id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("organizations.organization_id", ondelete="RESTRICT"), nullable=False, index=True)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    previous_knowledge_pack_version_id: Mapped[UUID | None] = mapped_column(Uuid(as_uuid=True), ForeignKey("knowledge_pack_versions.knowledge_pack_version_id", ondelete="RESTRICT"))
    mapping_version_ids: Mapped[list[str]] = mapped_column(JSONB, nullable=False)
    published_by: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), ForeignKey("users.user_id", ondelete="RESTRICT"), nullable=False)
    published_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=utc_now)
    schema_version: Mapped[str] = mapped_column(String(32), nullable=False, default="1.0.0", server_default="1.0.0")

File: frontend/src/shared/types/training.ts
export type ReviewStatus = 'open' | 'under_review' | 'mapped' | 'dismissed' | 'not_actionable'
export type MappingStatus = 'suggested' | 'draft' | 'testing' | 'approved' | 'published' | 'superseded' | 'rejected'
export type ValidationFamily = 'positive' | 'alternate_values' | 'negative' | 'wrong_scope' | 'negation' | 'conflict' | 'regression'

export interface UnresolvedBlock {
  unresolved_block_id: string
  audit_id: string
  device_id: string
  snapshot_id: string
  profile_id: string | null
  profile_version_id: string | null
  source_ir_node_ids: string[]
  evidence_refs: Record<string, unknown>[]
  raw_text: string
  surrounding_context: string
  unknown_reason: string
  candidate_field_ids: string[]
  affected_rule_ids: string[]
  fingerprint: string
  occurrence: Record<string, unknown>
  review_status: ReviewStatus
  assigned_mapping_version_id: string | null
  created_at: string
  updated_at: string
  schema_version: string
}

export interface CanonicalField {
  field_id: string
  expected_types: string[]
  allowed_scope_types: string[]
  domain: string
  description: string
  repeatable: boolean
}

export interface MappingDefinition {
  profile_applicability: { profile_ids?: string[]; profile_version_ids: string[] }
  structural_match: Record<string, unknown>
  target_field_id: string
  value_extraction: Record<string, unknown>
  unit_conversion: { operation: string }
  scope_resolution: { strategy: string }
  negation_behavior: { operation: string }
  removal_behavior: { operation: string }
  default_behavior: { operation: string }
  examples: Array<{ family: ValidationFamily; node: Record<string, unknown>; expected_match: boolean; expected_value?: unknown }>
}

export interface MappingVersion extends Omit<MappingDefinition, 'profile_applicability'> {
  mapping_version_id: string
  mapping_id: string
  mapping_key: string
  version: number
  previous_mapping_version_id: string | null
  title: string
  description: string
  status: MappingStatus
  profile_applicability: { profile_ids?: string[]; profile_version_ids: string[] }
  validation_results: Record<string, unknown> & { passed?: boolean; families?: Partial<Record<ValidationFamily, boolean>> }
  origin: 'built_in' | 'administrator' | 'ai_assisted'
  ai_suggestion_metadata: { confidence?: number; similar_mapping_refs?: string[]; provider?: Record<string, string> } | null
  created_by: string | null
  approved_by: string | null
  knowledge_pack_version_id: string | null
  created_at: string
  approved_at: string | null
  published_at: string | null
  schema_version: string
}

export interface ValidationRequest { validation_run_id: string; job_id: string; status: string }
export interface Publication { mapping: MappingVersion; knowledge_pack_version_id: string; knowledge_pack_version: number }
export interface ImpactAnalysis {
  mapping_version_id: string
  matching_unresolved_block_ids: string[]
  affected_device_ids: string[]
  affected_historical_audit_ids: string[]
  potentially_affected_unknown_findings: number
  historical_profile_audit_count: number
  creates_audit_revision: false
}
export interface KnowledgePack { knowledge_pack_id: string; pack_key: string; name: string; created_at: string }
export interface KnowledgePackVersion { knowledge_pack_version_id: string; knowledge_pack_id: string; version: number; previous_knowledge_pack_version_id: string | null; mapping_version_ids: string[]; published_by: string; published_at: string }
export interface TrainingProfile { profile_id: string; profile_version_id: string; vendor: string; product_family: string; os: string; reader: string | null; coverage: Record<string, unknown>; capabilities: string[] }

export interface MappingSuggestionPreview {
  suggestion: { definition: MappingDefinition; description: string; confidence: number; caveats: string[]; provider_metadata: Record<string, string> }
  evidence_references: Record<string, string>[]
  redaction_occurred: boolean
  truncation_occurred: boolean
  adoption_token: string
  evidence: Record<string, string>
  validation: string
}
export interface LocalAIStatus { provider: string; model: string; enabled: boolean; available: boolean; reason: string }
